resharper/src/resharper-unity/CSharp/Feature/Services/QuickFixes/PreferAddressByIdToGraphicsParamsQuickFix.cs

```diff
@@ -0,0 +1,258 @@
+            myArgument = warning.Argument;
+            myFieldName = myGraphicsPropertyName = warning.Literal;
+            if (!ValidityChecker.IsValidIdentifier(myFieldName))
+            {
+                myFieldName = "Property";
+            }
+
+            myArgumentExpression = warning.ArgumentExpression;
+            myMapFunction = warning.MapFunction;
+            myTypeName = warning.TypeName;
+        }
+        
+        protected override Action<ITextControl> ExecutePsiTransaction(ISolution solution, IProgressIndicator progress)
+        {
+            var factory = CSharpElementFactory.GetInstance(myInvocationExpression);
+
+            //  try find declaration where string to id conversation is done. If we don't find it, create in top-level class
+            var idDeclaration = TryFindDeclaration(myInvocationExpression, myGraphicsPropertyName, myTypeName, myMapFunction, out var name); 
+            if (idDeclaration == null)
+            {
+                var psiModule = myInvocationExpression.PsiModule;
+                var fieldInitializerValue = factory.CreateExpression("$0.$1($2)", 
+                    TypeFactory.CreateTypeByCLRName(myTypeName, myInvocationExpression.PsiModule),
+                    myMapFunction,
+                    myArgumentExpression.Copy());
+                
+                name = GetUniqueName(myInvocationExpression, myFieldName).NotNull();
+                var newDeclaration = factory.CreateFieldDeclaration(psiModule.GetPredefinedType().Int, name);
+                idDeclaration = newDeclaration.DeclaredElement;
+                if (idDeclaration == null)
+                    return null;
+                
+                newDeclaration.SetReadonly(true);
+                newDeclaration.SetStatic(true);
+                newDeclaration.SetAccessRights(AccessRights.PRIVATE);
+                newDeclaration.SetInitial(factory.CreateExpressionInitializer(fieldInitializerValue));
+
+                // TODO: [C#8] default interface implementations
+                // interface is not good place to add field declaration
+                var classDeclaration = GetTopLevelClassLikeDeclaration(myInvocationExpression);
+                classDeclaration.AddClassMemberDeclaration(newDeclaration);
+            }
+
+            // replace argument
+            var referenceExpression = factory.CreateReferenceExpression("$0", idDeclaration);
+            
+            
+            var argument = factory.CreateArgument(ParameterKind.VALUE, myArgument.NameIdentifier?.Name, referenceExpression);
+            myArgument.ReplaceBy(argument);
+            return null;
+        } 
+
+        private static IDeclaredElement TryFindIdDeclarationInClassLikeDeclaration(IClassLikeDeclaration declaration, string propertyName,
+            string requiredTypeName, string requiredMethodName, out string name)
+        {
+            name = null;
+            foreach (var member in declaration.ClassMemberDeclarations)
```

Comment:
This code only checks declarations syntaxically located in the passed declaration.
I belive that my original example still won't reuse existing property, could you please add a test for it?
```cs
public partial class TestClass
{
  public void Method(Material material)
  {
    material.SetFloat("te{caret}st", 10.0f);
  }
}

public partial class TestClass
{
  private static readonly int Test = Shader.PropertyToID("test");
}
```

You can check all declarations of a type like this
```cs
foreach (var declarationPart in declaration.DeclaredElement.GetDeclarations())
```